pkg/controller/job/job_controller.go

```diff
@@ -772,6 +772,29 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	}
 
 	activePods := controller.FilterActivePods(pods)
+
+	// As part of KEP-3939,
+	// we will allow for opt-in control of including terminating pods
+	// in our activePod list.
+	// This works because we pass this list into managePods which counts the list of activePods.
+	var terminatingPods int32
+	if feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) {
+		terminatingPods = controller.CountTerminatingPods(pods)
+		// Since defaulting is not allowed for new fields we are going to make some assumptions
+		// If PodReplacementPolicy is nil and PodFailurePolicy is not set then default will be TerminatingOrFailed
+		// This means that we wont include terminating pods in the count.
+		includeTerminating := false
+		// If PodFailurePolicy is enabled we will assume Failed.  Any other value is a validation error.
+		if job.Spec.PodReplacementPolicy == nil && job.Spec.PodFailurePolicy != nil {
+			includeTerminating = true
+		} else if job.Spec.PodReplacementPolicy != nil && *job.Spec.PodReplacementPolicy == batch.Failed {
+			includeTerminating = true
+		}
+		if includeTerminating && terminatingPods > 0 {
+			terminatingPodsList := controller.FilterTerminatingPods(pods)
+			activePods = append(activePods, terminatingPodsList...)
+		}
+	}
 	active := int32(len(activePods))
```

Comment:
Can we please do without changing the meaning of active pods from the API perspective as discussed before? https://github.com/kubernetes/enhancements/pull/3940#discussion_r1195595365

Reply:
I find it is justifiable to count terminating pods as active when the replacement policy is `Failed` . 

It is consistent with the note `Users can have a simpler mental model for the status.active field as a number of pods doing useful work.`, because terminating pods still can do useful work. They have a grace period (30s or more) they can use to complete computations and end up in the `Succeeded` phase.

@atiratree @alculquicondor wdyt?

Reply:
I suppose, it can be, depending on how you define useful work. Although mostly applications should just try to cleanup. 

I think the main issue is that you have to read spec to interpret the status. And yes minimizing the API changes is a plus too.

Reply:
Makes sense, let's don't change the `activePods` and return terminating pods separately. This means we need to combine activePods with terminating in some places in the code, I guess this can be useful to easily pass down a new field `terminatingPods` if needed: https://github.com/kubernetes/kubernetes/pull/119193

Reply:
I think this may be a larger change than we should do for an alpha feature.  `manageJobs` passes in active pods to at least 2-3 functions.  I went the route of including this list because I didn't want to change a lot of code for an alpha feature.  

I can go with creating a separate list of active and terminating and adding logic to handle both these lists in positive/negative cases of the feature toggle.  If we go this route I think we may need to differ this feature until 1.29 as I think manageJobs may need some changes to handle both active and terminating lists separately.  I have to do some research into how functions like `activePodsForRemoval` would need to change if I wanted to pass in a terminating list also.

Reply:
IIUC the rename does not fully address the comment @atiratree, about what is exposed via API. In particular you have expose this count via API: https://github.com/kubernetes/kubernetes/blob/f0464c6f978a677f29534f870cbcacc56a4c225c/pkg/controller/job/job_controller.go#L931

Thus it changes the counts for active. Now, it is debatable if we want that or not - cause as I pointed out under some definition of active, as doing useful work, one can imaging a terminating pod still completing its useful work within grace period. 

However, I agree this is rather exceptional, and so leaning now towards not including terminating in the counter for active, but not strong opinion.

IIUC if we go this path, then on the technical level we want to have two lists of pods `activePods` as currently and `terminatingPods`. In most places when we iterate over activePods we would include the terminating pods in the iteration. However, when deleting pods I don't see a need to pass already terminating pods here: https://github.com/kubernetes/kubernetes/blob/f0464c6f978a677f29534f870cbcacc56a4c225c/pkg/controller/job/job_controller.go#L1454C18-L1454C38

Reply:
wdyt @alculquicondor ?

Reply:
We discussed this offline with @kannon92 and @alculquicondor, and the verdict is to keep the `job.Status.active` as currently, ie. terminating pods should not be included. 

At the code level we will need to maintain `activePods` as is. Also, we will need to have `terminatingPods` used in `manageJob` to do not recreate indexes corresponding to terminatingPods when `podRecreationPolicy: Failed`. I think the list will only be needed for indexed jobs, it seems for regular job the count is enough.

Reply:
Pushed up some changes to reflect this but working on a bug currently.